### PR TITLE
Add support for reading Parquet logical type timestamp with millis and micros units. #743

### PR DESCRIPTION
**Pull request by jcferretti**
We need both page support for them and small schema loading changes.  The code is coming from work that Paul and Ryan did in the DHE `pjc_IDS-7514` branch.

### CODE REVIEW

File: DB/src/main/java/io/deephaven/db/v2/locations/parquet/topage/ToDBDateTimePage.java [hunk 1]
```diff
@@ -3,39 +3,77 @@
...
 
-    public static <ATTR extends Attributes.Any> ToDBDateTimePage<ATTR> create(@NotNull Class<?> nativeType) {
+    private static <ATTR extends Attributes.Any> ToDBDateTimePage<ATTR> fromUnitToNanos(final LongUnaryOperator convertUnitToNanos) {
+        return new ToDBDateTimePage<ATTR>() {
+            private long toNanos(final long v) {
+                return convertUnitToNanos.applyAsLong(v);
+            }
+
+            @Override
+            public final long[] convertResult(@NotNull final Object result) {
```
Comment (rcaudy) [inline DB/src/main/java/io/deephaven/db/v2/locations/parquet/topage/ToDBDateTimePage.java:22]: This is definitely not the design we want. You don't need or want to override `convertResult` in the nanos case; array copy is the right approach in that scenario.

File: DB/src/main/java/io/deephaven/db/v2/locations/parquet/topage/ToDBDateTimePage.java [hunk 2]
```diff
@@ -3,39 +3,77 @@
...
+import java.util.function.LongUnaryOperator;
 
-    private static ToDBDateTimePage INSTANCE = new ToDBDateTimePage<>();
+public abstract class ToDBDateTimePage<ATTR extends Attributes.Any> extends ToLongPage<ATTR> {
 
-    public static <ATTR extends Attributes.Any> ToDBDateTimePage<ATTR> create(@NotNull Class<?> nativeType) {
+    private static <ATTR extends Attributes.Any> ToDBDateTimePage<ATTR> fromUnitToNanos(final LongUnaryOperator convertUnitToNanos) {
+        return new ToDBDateTimePage<ATTR>() {
+            private long toNanos(final long v) {
+                return convertUnitToNanos.applyAsLong(v);
```
Comment (rcaudy) [inline DB/src/main/java/io/deephaven/db/v2/locations/parquet/topage/ToDBDateTimePage.java:18]: I know you were doing this to appeal to my sensibilities, but I suspect the abstract method is slightly better/more natural.

Comment (rcaudy) [reply DB/src/main/java/io/deephaven/db/v2/locations/parquet/topage/ToDBDateTimePage.java:18]: I don't love the anonymous inner class approach. I think the best design is to add an abstract `toDateTime` method as you did before, and use it in `toDbArray` for all `ToDBDateTimePage` impls, and use it in an overriden `convertResult` in the millis and micros implementations. What you had before this round was better, just needed to add the `convertResult` overrides.

File: DB/src/main/java/io/deephaven/db/v2/locations/parquet/topage/ToDBDateTimePage.java [hunk 3]
```diff
@@ -3,39 +3,77 @@
...
+                return resultLongs;
+            }
+
+            @Override
+            @NotNull
+            public DbArray<DBDateTime> makeDbArray(long[] result) {
+                DBDateTime[] to = new DBDateTime[result.length];
+
+                for (int i = 0; i < result.length; ++i) {
+                    to[i] = new DBDateTime(toNanos(result[i]));
```
Comment (rcaudy) [inline DB/src/main/java/io/deephaven/db/v2/locations/parquet/topage/ToDBDateTimePage.java:39]: This is wrong. You aren't mapping null times to the null `DBDateTime` correctly. The previous impl was correct.

**Review by rcaudy — APPROVED** [review]

**Comment by rcaudy** [comment]
Link to #294

**Comment by jcferretti** [comment]
Fixed by https://github.com/deephaven/deephaven-core/pull/744

**Pull request merged**